src/window-rules.c

```diff
@@ -0,0 +1,83 @@
+// SPDX-License-Identifier: GPL-2.0-only
+#define _POSIX_C_SOURCE 200809L
+#include "config.h"
+#include <assert.h>
+#include <cairo.h>
+#include <glib.h>
+#include <strings.h>
+#include <wlr/util/log.h>
+#include "action.h"
+#include "common/match.h"
+#include "config/rcxml.h"
+#include "labwc.h"
+#include "view.h"
+#include "window-rules.h"
+
+static const char *
+get_identifier(struct view *view)
+{
+	switch (view->type) {
+	case LAB_XDG_SHELL_VIEW:
+		return view_get_string_prop(view, "app_id");
+#if HAVE_XWAYLAND
+	case LAB_XWAYLAND_VIEW:
+		return view_get_string_prop(view, "class");
+#endif
+	}
+	return "";
+}
+
+void
+window_rules_apply(struct view *view, enum window_rule_event event)
+{
+	const char *identifier = get_identifier(view);
+	if (!identifier) {
+		return;
+	}
+
+	struct window_rule *rule;
+	wl_list_for_each(rule, &rc.window_rules, link) {
+		if (!rule->identifier || rule->event != event) {
+			continue;
+		}
+		if (match(rule->identifier, identifier)) {
+			actions_run(view, view->server, &rule->actions, 0);
+		}
+	}
+}
+
+enum property
+window_rules_get_property(struct view *view, const char *property)
+{
+	assert(property);
+
+	const char *identifier = get_identifier(view);
+	if (!identifier) {
+		goto out;
+	}
+
+	/*
+	 * We iterate in reverse here because later items in list have higher
+	 * priority. For example, in the config below we want the return value
+	 * for foot's "serverDecoration" property to be "unset".
+	 *
+	 *     <windowRules>
+	 *       <windowRule identifier="*" serverDecoration="no"/>
+	 *       <windowRule identifier="foot" serverDecoration="unset"/>
+	 *     </windowRules>
+	 */
+	struct window_rule *rule;
+	wl_list_for_each_reverse(rule, &rc.window_rules, link) {
+		if (!rule->identifier) {
+			continue;
+		}
+		if (match(rule->identifier, identifier)) {
+			if (!strcasecmp(property, "serverDecoration")) {
+				return rule->server_decoration;
+			}
+			wlr_log(WLR_ERROR, "unknown property '%s'", property);
+		}
+	}
```

Comment:
I think we should `continue` when the requested value is set to `LAB_PROP_UNSPECIFIED`. Otherwise this would return `LAB_PROP_UNSPECIFIED` if asked for the `serverDecorations` of `foo.bar`:
```xml
<windowRules>
    <windowRule identifier="*" serverDecorations="no" />
    <windowRule identifier="foo.bar" />
```

Reply:
Good catch indeed.
I've pushed a fix which differentiates between unspecified and unset.